cli/cppcheckexecutor.cpp

```diff
@@ -71,6 +72,157 @@
+        }
+
+        picojson::value serializeRuns(const std::string& productName, const std::string& version) const {
+            picojson::object driver;
+            driver["name"] = picojson::value(productName);
+            driver["version"] = picojson::value(version);
+            driver["informationUri"] = picojson::value("https://cppcheck.sourceforge.io");
+            driver["rules"] = picojson::value(serializeRules());
+            picojson::object tool;
+            tool["driver"] = picojson::value(driver);
+            picojson::object run;
+            run["tool"] = picojson::value(tool);
+            run["results"] = picojson::value(serializeResults());
+            picojson::array runs{picojson::value(run)};
+            return picojson::value(runs);
+        }
+
+        std::string serialize(std::string productName) const {
+            const auto nameAndVersion = Settings::getNameAndVersion(productName);
+            productName = nameAndVersion.first.empty() ? "Cppcheck" : nameAndVersion.first;
+            std::string version = nameAndVersion.first.empty() ? CppCheck::version() : nameAndVersion.second;
+            if (version.find(' ') != std::string::npos)
+                version.erase(version.find(' '), std::string::npos);
+
+            picojson::object doc;
+            doc["version"] = picojson::value("2.1.0");
+            doc["$schema"] = picojson::value("https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json");
+            doc["runs"] = serializeRuns(productName, version);
+
+            return picojson::value(doc).serialize(true);
+        }
+    private:
+
+        static std::string sarifSeverity(const ErrorMessage& errmsg) {
+            if (ErrorLogger::isCriticalErrorId(errmsg.id))
+                return "error";
+            switch (errmsg.severity) {
+            case Severity::error:
+            case Severity::warning:
+            case Severity::style:
+            case Severity::portability:
+            case Severity::performance:
+                return "warning";
+            case Severity::information:
+            case Severity::internal:
+            case Severity::debug:
+            case Severity::none:
+                return "note";
+            }
+            return "note";
+        }
+
+        static std::string sarifPrecision(const ErrorMessage& errmsg) {
+            if (errmsg.certainty == Certainty::inconclusive)
+                return "normal";
+            return "high";
+        }
+
+        static std::string getHash(const ErrorMessage& errmsg) {
+            const std::string s = errmsg.toString(false, "{file}:{line}:{column}: {message} {id} {code}", "{file}:{line}:{column} {info} {code}");
+            std::ostringstream os;
+            //std::cout << s << std::endl;
+            os << std::hex << std::hash<std::string> {}(s);
+            return os.str();
+        }
+
+        std::vector<ErrorMessage> mFindings;
+    };
```

Comment:
This is fine. Since it includes the line and column number, I'm not sure how "resistant" it is to small changes. A better idea would be to hash the source token(s). That way, if the code changes a bit (moves up/down a line, indented, etc), it would still get recognized as the same finding. But I understand that might be a lot of work.

Another option: you can omit the `partialFingerprints` property from the SARIF file altogether, and rely on the GitHub Action [`upload-sarif`](https://github.com/github/codeql-action) to generate the `partialFingerprints` and insert them into the SARIF file on upload to GitHub. The drawback is that the `partialFingerprints` property would only be added when cppcheck is run in GitHub Actions. If you generated a Cppcheck SARIF file outside of GitHub Actions and then uploaded it to GitHub, you might get duplicate alerts.

Reply:
yes I think this is quite tricky. I'll remove the partialFingerprints then.